lib/src/editorhelper.cpp

```diff
@@ -234,4 +235,255 @@ std::unique_ptr<uhdr_raw_image_ext_t> apply_resize(uhdr_raw_image_t* src, int ds
+    for (size_t i = 0; i < dst->h; i += 2) {
+      for (size_t j = 0; j < dst->w; j += 2) {
+        float r0 = float(rgbData[srcStride * i + j] & 0xff);
+        float g0 = float((rgbData[srcStride * i + j] >> 8) & 0xff);
+        float b0 = float((rgbData[srcStride * i + j] >> 16) & 0xff);
+
+        float r1 = float(rgbData[srcStride * i + (j + 1)] & 0xff);
+        float g1 = float((rgbData[srcStride * i + (j + 1)] >> 8) & 0xff);
+        float b1 = float((rgbData[srcStride * i + (j + 1)] >> 16) & 0xff);
+
+        float r2 = float(rgbData[srcStride * (i + 1) + j] & 0xff);
+        float g2 = float((rgbData[srcStride * (i + 1) + j] >> 8) & 0xff);
+        float b2 = float((rgbData[srcStride * (i + 1) + j] >> 16) & 0xff);
+
+        float r3 = float(rgbData[srcStride * (i + 1) + (j + 1)] & 0xff);
+        float g3 = float((rgbData[srcStride * (i + 1) + (j + 1)] >> 8) & 0xff);
+        float b3 = float((rgbData[srcStride * (i + 1) + (j + 1)] >> 16) & 0xff);
+
+        r0 /= 255.0f;
+        g0 /= 255.0f;
+        b0 /= 255.0f;
+
+        r1 /= 255.0f;
+        g1 /= 255.0f;
+        b1 /= 255.0f;
+
+        r2 /= 255.0f;
+        g2 /= 255.0f;
+        b2 /= 255.0f;
+
+        r3 /= 255.0f;
+        g3 /= 255.0f;
+        b3 /= 255.0f;
+
+        float y0 = coeffs[0] * r0 + coeffs[1] * g0 + coeffs[2] * b0;
+        float y1 = coeffs[0] * r1 + coeffs[1] * g1 + coeffs[2] * b1;
+        float y2 = coeffs[0] * r2 + coeffs[1] * g2 + coeffs[2] * b2;
+        float y3 = coeffs[0] * r3 + coeffs[1] * g3 + coeffs[2] * b3;
+
+        float u0 = (b0 - y0) / coeffs[3];
+        float u1 = (b1 - y1) / coeffs[3];
+        float u2 = (b2 - y2) / coeffs[3];
+        float u3 = (b3 - y3) / coeffs[3];
+
+        float v0 = (r0 - y0) / coeffs[4];
+        float v1 = (r1 - y1) / coeffs[4];
+        float v2 = (r2 - y2) / coeffs[4];
+        float v3 = (r3 - y3) / coeffs[4];
+
+        y0 = y0 * 255.0f + 0.5f;
+        y0 = CLIP3(y0, 0.0f, 255.0f);
+        yData[dst->stride[UHDR_PLANE_Y] * i + j] = uint8_t(y0);
+
+        y1 = y1 * 255.0f + 0.5f;
+        y1 = CLIP3(y1, 0.0f, 255.0f);
+        yData[dst->stride[UHDR_PLANE_Y] * i + j + 1] = uint8_t(y1);
+
+        y2 = y2 * 255.0f + 0.5f;
+        y2 = CLIP3(y2, 0.0f, 255.0f);
+        yData[dst->stride[UHDR_PLANE_Y] * (i + 1) + j] = uint8_t(y2);
+
+        y3 = y3 * 255.0f + 0.5f;
+        y3 = CLIP3(y3, 0.0f, 255.0f);
+        yData[dst->stride[UHDR_PLANE_Y] * (i + 1) + j + 1] = uint8_t(y3);
+
+        u0 = (u0 + u1 + u2 + u3) / 4;
+        v0 = (v0 + v1 + v2 + v3) / 4;
+
+        u0 = u0 * 255.0f + 0.5 + 128.0f;
+        v0 = v0 * 255.0f + 0.5 + 128.0f;
+
+        u0 = CLIP3(u0, 0.0f, 255.0f);
+        v0 = CLIP3(v0, 0.0f, 255.0f);
+
+        uData[dst->stride[UHDR_PLANE_U] * (i / 2) + (j / 2)] = uint8_t(u0);
+        vData[dst->stride[UHDR_PLANE_V] * (i / 2) + (j / 2)] = uint8_t(v0);
+      }
+    }
+  } else if (src->fmt == UHDR_IMG_FMT_12bppYCbCr420) {
```

Comment:
Question why do we need that?

Ideally I'm thinking this method is converting RGB input to YUV output, and if user calls this method with YUV input, we don't need to do anything but tell the user this is wrong call. With this code I think it will potentially grow into a bigger TODO like YUV 444 transfer to YUV 420, and we don't have support for that currently. So I would suggest remove this and the following if clause, just keep RGB in and YUV out.

Reply:
For yuv420 and p010 formats, the input is copied from source buffer to a buffer internal to the uhdr library. uhdr_enc_set_raw_image() receives a raw buffer to encode. Once the call returns it is unclear if the application can free the memory or keep it alive through the encode cycle. So currently we are doing a copy to an internal buffer and once the call returns the input can be freed or reused for other purpose. This copy code is moved from ultrahdr_api.cpp to editor_helper.cpp